### PR TITLE
Bug 1244169 - Implement Tabs sync data adapter.

### PR DESCRIPTION
**Pull request by weilonge**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4395b9cb07fbec87cae4d8ecc8e438d68e2eedb7

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=025931469a367535ebc97bf549aa2e603914a021

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7c402628d31e41205957969d077a0333a760b525

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f514775bf3243e708783adcbd23bbece128a1bf9

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=396facfde21abce2e37f9902da8e3b39241f73fb

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=baf96bcbdfaecc12b897ae35b605904c5a858b3d

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=13a9b6d87f08f366f8ed7325b69d0bfd28efe406

File: apps/sync/index.html [hunk 1]
```diff
@@ -19,4 +19,5 @@
   <script src="shared/js/lazy_loader.js"></script>
   <script src="shared/js/iac_handler.js"></script>
   <script src="shared/js/sync/bootstrap.js"></script>
+  <script src="js/scratchpad.js"></script>
```
Comment (michielbdejong) [inline apps/sync/index.html]: Added by accident? If that file is needed, give it a more descriptive name.

File: apps/sync/manifest.webapp [hunk 2]
```diff
@@ -2,7 +2,6 @@
   "name": "Sync",
   "description": "Syncs data with Firefox Sync",
   "type": "certified",
-  "role": "system",
```
Comment (michielbdejong) [inline apps/sync/manifest.webapp]: wasn't this needed to make the sync app run in the background?

File: build/config/phone/custom-prefs.js [hunk 3]
```diff
@@ -2,3 +2,6 @@
 user_pref('devtools.responsiveUI.customWidth', 320);
 user_pref('devtools.responsiveUI.customHeight', 480);
 user_pref('devtools.responsiveUI.currentPreset', 'custom');
+user_pref('devtools.debugger.remote-enabled', true);
+user_pref('devtools.chrome.enabled', true);
```
Comment (michielbdejong) [inline build/config/phone/custom-prefs.js]: Added by accident?

File: shared/js/sync/bootstrap.js [hunk 4]
```diff
@@ -18,6 +18,8 @@
   DataAdapters
 */
 
+var SYNC_DEBUG = {};
+
```
Comment (michielbdejong) [inline shared/js/sync/bootstrap.js]: Added by accident?

File: shared/js/sync/bootstrap.js [hunk 5]
```diff
@@ -127,7 +130,7 @@ const Bootstrap = (() => {
           sendPortMessage({
             id: request.id
           });
-          window.close();
+          //window.close();
```
Comment (michielbdejong) [inline shared/js/sync/bootstrap.js]: Removed by accident?

File: shared/js/sync/bootstrap.js [hunk 6]
```diff
@@ -90,6 +92,7 @@ const Bootstrap = (() => {
         assertion: request.assertion,
         adapters: DataAdapters
       });
+      SYNC_DEBUG.syncEngine = syncEngine;
```
Comment (michielbdejong) [inline shared/js/sync/bootstrap.js]: Added by accident?

**Comment by michielbdejong** [comment]
In general, it's a pity that we now have so much copy-paste duplicated code, but I guess it's a separate goal to refactor that one day.

File: apps/sync/index.html (hunk 1 above)
Comment (weilonge) [inline apps/sync/index.html]: This is only for debugging purpose. Sorry for the misunderstanding.
Please see the commit which will be landed to master:
[https://github.com/weilonge/gaia/commit/13a9b6d87f08f366f8ed7325b69d0bfd28efe406]

File: apps/sync/manifest.webapp (hunk 2 above)
Comment (weilonge) [inline apps/sync/manifest.webapp]: This is only for debugging purpose.

File: build/config/phone/custom-prefs.js (hunk 3 above)
Comment (weilonge) [inline build/config/phone/custom-prefs.js]: This is only for debugging purpose.

File: shared/js/sync/bootstrap.js (hunk 4 above)
Comment (weilonge) [inline shared/js/sync/bootstrap.js]: This is only for debugging purpose.

File: shared/js/sync/bootstrap.js (hunk 5 above)
Comment (weilonge) [inline shared/js/sync/bootstrap.js]: This is only for debugging purpose.

File: shared/js/sync/bootstrap.js (hunk 6 above)
Comment (weilonge) [inline shared/js/sync/bootstrap.js]: This is only for debugging purpose.

**Comment by weilonge** [comment]
@michielbdejong Thanks for your comments! I still leave the debugging purpose code in this PR because that's easier for development. Sorry for the confusion.
I feel the same thing that the duplicate codes should be refactored. Did you create a bug to refactor this? If not, I will create one for this.

**Comment by michielbdejong** [comment]
Yes, it's https://bugzilla.mozilla.org/show_bug.cgi?id=1224183

File: apps/sync/index.html (hunk 1 above)
Comment (michielbdejong) [inline apps/sync/index.html]: Ah right, there are two commits in this PR, sorry, got it :)

**Comment by michielbdejong** [comment]
@weilonge To review this, I search-and-replace 'tab' to 'bookmark' in `adapters/tabs.js`, and checked if it does the same as the bookmarks adapter that's already in master. Differences I found:

* You're not using `getDataStoreId`/`setDataStoreId` - this is because you're using `dataStoreId === fxSyncId`, right? Will this be ok when you want to display the open tabs in the UI - don't they need to be sorted somehow?
* How does this relate to the `createdLocally` mechanism? I think open tabs are per-device, right? So entries in the tab collection that are created locally will already be recognizable by their `clientName` field?
* Speaking of which, why have a `mergeRecordsToDataStore` function at all? I think you can just store all incoming records as-is, and push all outgoing records as-is, with the TV's `clientName`.

Given that the adapters are in need of a refactor anyway, I think all these are issues you can fix during that refactor, in a follow-up issue if you want.

I didn't review the tests in detail, they look like they're quite similar to the tests for the other adapters, and that's also code that would change again anyway if you refactor the adapters.

There are quite a few failing tests on TreeHerder, most of them intermittent and probably unrelated, right?

So apart from removing the debug-commit from the PR, r+ from me - with the follow-up ticket for the adapters refactor.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ff6fd515df22a77b69181115b63df421c45cfc20

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9264cd966d91f308054307a5dbc116e0e6b6c736

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=efb5962fa4cda9e3db02f9185a42354ebdcb8240

**Pull request merged**